hilti/runtime/include/util.h

```diff
@@ -285,7 +286,48 @@ constexpr auto enumerate(T&& iterable) {
  * @param str string to expand
  * @return A UTF8 string with escape sequences expanded
  */
-std::string expandEscapes(std::string s);
+std::string expandUTF8Escapes(std::string s);
+
+namespace render_style {
+
+/**
+ * Flags specifying escaping style when rendering raw data for printing. The
+ * default style renders all non-printable characters as hex escapes (`\xNN`)
+ * and escapes backslashes with a second backslash. Any specified flags modify
+ * the default style accordingly.
+ */
+enum class Bytes {
+    Default = 0,                    /**< name for unmodified default style */
+    EscapeQuotes = (1U << 1U),      /**< escape double quotes with backslashes */
+    UseOctal = (1U << 2U),          /**< escape non-printables with `\NNN` instead of `\xNN` */
+    NoEscapeBackslash = (1U << 3U), /**< do not escape backslashes */
+};
+
+/**
+ * Flags specifying escaping style when rendering UTF8 strings for printing.
+ * The default style escapes control characters and null bytes with
+ * corresponding C-style control escapes (e.g., `\n`, `\0`), and escapes any
+ * backslashes with a second backslash. Any specified flags modify the default
+ * style accordingly. If not otherwise noted, any escapings are reversible
+ * through `expandUTF8Escapes()`.
+ */
+enum class UTF8 {
+    Default = 0,                    /**< name for unmodified default style */
+    EscapeQuotes = (1U << 1U),      /**< escape double quotes with backslashes */
+    NoEscapeBackslash = (1U << 2U), /**< do not escape backslashes; this may leave the result non-reversible */
+    NoEscapeControl = (1U << 3U),   /**< do not escape control characters and null bytes */
+    NoEscapeHex =
+        (1U << 4U), /**< do not escape already existing `\xNN` escape codes; this may leave the result non-reversible */
```

Comment:
same comment as before, I'd prefer these flags be changed to `EscapeBackslash`, `EscapeControl`, and `EscapeHex` - escaping features that you opt into rather than a mix.